libopenage/pathfinding/cost_field.cpp

```diff
@@ -33,29 +34,42 @@ cost_t CostField::get_cost(size_t idx) const {
 	return this->cells.at(idx);
 }
 
-void CostField::set_cost(const coord::tile_delta &pos, cost_t cost) {
+void CostField::set_cost(const coord::tile_delta &pos, cost_t cost, time::time_t &changed) {
 	this->cells[pos.ne + pos.se * this->size] = cost;
+	this->changed = changed;
 }
 
-void CostField::set_cost(size_t x, size_t y, cost_t cost) {
+void CostField::set_cost(size_t x, size_t y, cost_t cost, time::time_t &changed) {
 	this->cells[x + y * this->size] = cost;
+	this->changed = changed;
 }
 
 void CostField::set_cost(size_t idx, cost_t cost) {
 	this->cells[idx] = cost;
+	this->changed = time::TIME_ZERO;
+}
+
+void CostField::set_cost(size_t idx, cost_t cost,time::time_t &changed) {
+	this->cells[idx] = cost;
+	this->changed = changed;
 }
 
 const std::vector<cost_t> &CostField::get_costs() const {
 	return this->cells;
 }
 
-void CostField::set_costs(std::vector<cost_t> &&cells) {
+void CostField::set_costs(std::vector<cost_t> &&cells, time::time_t &changed) {
 	ENSURE(cells.size() == this->cells.size(),
 	       "cells vector has wrong size: " << cells.size()
 	                                       << "; expected: "
 	                                       << this->cells.size());
 
 	this->cells = std::move(cells);
+	this->changed = changed;
+}
+
+bool CostField::is_dirty(time::time_t &time) {
+	return time <= this->changed;
```

Comment:
That makes sense. I was thinking from the perspective of the cost field. If a time is passed in _before_ the latest change from the cost field's perspective, it would be an old value, e.g. dirty. But we oughtn't care about the cost field's feelings here, correct, only the caller's?

Reply:
Let's say two requests are sent to a cost field after it is updated at time T: one at T + 1, and the other at time T + 3. Using (`>=`), the cost field returns dirty if the request happens after the cost field is changed. If both requests are from the integrator, then the integrator will be told that the cost field is dirty at time T + 1, and then again at time T + 3. But the cost field was only updated once. So with this approach, the cost field will be dirty for every request after time T.

Let's say that the cost field is called by the integrator 20 times after it was changed (Let's say 20 commands were given in a sector after a building was placed). That means the cache would be emptied 20 times!

So then is a time_t even the right thing to be using here?

Reply:
Ah I see where you are coming from. I think the confusion comes mostly from how I designed the issue, since it leaves out a few things that would make it a "not-so-good-first-issue" :D

So the general idea is that that the integration/flow fields must be recomputed when the cost field becomes dirty. So at any time _before_ it becomes dirty (`t` < `changed`) the cache can still be used, while at any time _after_ it becomes dirty (`t` >= `changed`) the cache should be recomputed (once). I think renaming `changed` to `valid_until` might make more clear that the dirtyness is linked to cache invalidation. 

But you're right that the cache would currently be evicted every time where `t` >= `valid_until`. What's currently missing is a mechanism to clear the "dirty" status from the cost field after the cache was updated. We could clear the dirty status by adding a method that sets `valid_until` to `time::TIME_MAX`. Then `is_dirty` would always fail until another change happens.

Reply:
Nice, an elegant solution that allows us to use time instead of boolean. This will come in handy when we add cost stamps and tracking, as time will already be getting involved with the pathfinding system thanks to this update.